Packages/ClientRuntime/Sources/Waiters/Waiter.swift

```diff
@@ -0,0 +1,103 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+/// An object used to wait on an operation.
+/// Intended to be a generic type for use when waiting on any Smithy `operation`.
+/// May be reused for multiple waits, including concurrent operations.
+public class Waiter<Input, Output> {
+
+    /// The configuration this waiter was created with.
+    public let config: WaiterConfig<Input, Output>
+
+    /// The operation that this waiter will call one or more times while waiting on the success condition.
+    public let operation: (Input) async throws -> Output
+
+    /// Allows for advancing of time during waiter unit tests.
+    /// The default hook does nothing.
+    var postSchedulerUpdateHook: (WaiterScheduler) -> Void = { _ in }
```

Comment:
is this for unit tests only? It is bad approach  if something is needed just for testing. It should be refactored.

Reply:
I can refactor tests to use dependency injection of a mocked-out `WaiterScheduler`, but it will add at least as much complexity as this does.  I normally prefer not to add properties like this that are for testing, but it seemed the best option here.

Reply:
I refactored the logic inside the `while` loop to a new type that can be tested separately.  It is dependency-injected by method parameter to allow a mock to be used for testing.